pkg/k8s/watchers/crd.go

```diff
@@ -0,0 +1,294 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (k *K8sWatcher) crdsInit(client apiextclientset.Interface) {
+	crds := newCRDState(
+		map[string]bool{
+			cnpCRD:  false,
+			ccnpCRD: false,
+			cepCRD:  false,
+			cnCRD:   false,
+			cidCRD:  false,
+			clrpCRD: false,
+		},
+	)
+
+	_, crdController := informer.NewInformer(
+		newListWatchFromClient(
+			newCRDGetter(client),
+			fields.Everything(),
+		),
+		&metav1.PartialObjectMetadata{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
+					crds.Lock()
+					crds.m["crd:"+pom.GetName()] = true
+					crds.Unlock()
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
+					crds.Lock()
+					crds.m["crd:"+pom.GetName()] = false
+					crds.Unlock()
+				}
+			},
+		},
+		nil,
+	)
+
+	// Create a context so that we can timeout after the configured CRD wait
+	// peroid.
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.CRDWaitTimeout)
+	for crd := range crds.m {
+		k.blockWaitGroupToSyncResources(
+			ctx.Done(),
+			nil,
+			func() bool {
+				crds.Lock()
+				defer crds.Unlock()
+				return crds.m[crd]
+			},
+			crd,
+		)
+	}
+
+	// The above loop will call blockWaitGroupToSyncResources to populate the
+	// K8sWatcher state with the current state of the CRDs. It will check the
+	// state of each CRD, with the function provided below. If the function
+	// reports that the given CRD is true (has been synced), it will close a
+	// channel associated with the given CRD. A subsequent call to
+	// (*K8sWatcher).WaitForCacheSync will notice that a given CRD's channel
+	// has been closed. Once all the CRDs passed to WaitForCacheSync have had
+	// their channels closed, the function unblocks.
+	//
+	// Meanwhile, the Goroutine below kicks off the controller that was
+	// instantiated above, and enters a loop looking for (1) if the context has
+	// deadlined or (2) if the entire CRD state has been synced (all CRDs found
+	// in the cluster). If (1) is encountered, then Cilium will fatal because
+	// it cannot proceed if the CRDs are not present. If (2) is encountered,
+	// then make sure the controller has exited by cancelling the context and
+	// we return out.
+
+	go func() {
+		defer cancel()
+		go crdController.Run(ctx.Done())
+		k.k8sAPIGroups.addAPI(k8sAPIGroupCRD)
+
+		for {
+			select {
+			case <-ctx.Done():
+				if err := ctx.Err(); err != nil {
+					log.WithError(err).
+						Fatalf("Unable to find all Cilium CRDs necessary within "+
+							"%v timeout. Please ensure that Cilium Operator is "+
+							"running, as it's responsible for registering all "+
+							"the Cilium CRDs.",
+							option.Config.CRDWaitTimeout)
+				}
+			case <-time.Tick(1 * time.Second):
+				if crds.isSynced() {
+					log.Info("All Cilium CRDs have been found and are available")
+					return
+				}
+			}
```

Comment:
Looking for feedback on if we can do this in a more efficient way. I could only think of polling...

Reply:
It's not the worst thing in the world to poll this every second. I guess a cleaner approach would be to have a channel to check whether all the CRDs were synced, which would be tied into `crdState`. Then the logic inside `isSynced()` would run somewhere in a separate goroutine to close the "CRDsSynced" channel when the current logic returns true.

Reply:
Thanks. Yeah it doesn't seem like it's possible to do this without polling in someway. Whether the polling happens directly in the loop or inside a helper goroutine somewhere else, the functionality is the same. I am keeping the current approach as adding another channel just adds extra complexity and indirection that doesn't provide much benefit, besides maybe "elegance". I like the directness of the code currently. :)

Reply:
Agree, I think that's the right tradeoff.